fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/values/MessageHelpers.java

```diff
@@ -174,6 +174,30 @@ public static Descriptors.FieldDescriptor findFieldDescriptorOnMessageByOrdinal(
         return message.getDescriptorForType().getFields().get(fieldOrdinal);
     }
 
+    public static boolean compareMessageEquals(@Nonnull MessageOrBuilder m1, @Nonnull MessageOrBuilder m2) {
+        if (m1.getDescriptorForType().getFields().size() != m2.getDescriptorForType().getFields().size()) {
+            return false;
+        }
+        // compare FieldDescriptors one by one
+        for (int i = 1; i <= m1.getDescriptorForType().getFields().size(); i++) {
+            Descriptors.FieldDescriptor f1 = findFieldDescriptorOnMessage(m1, i);
+            Descriptors.FieldDescriptor f2 = findFieldDescriptorOnMessage(m2, i);
+            // do not support repeated or nested fields in the message
+            if (f1.isRepeated() || f1.isMapField() || f1.getJavaType() == Descriptors.FieldDescriptor.JavaType.MESSAGE
```

Comment:
This should not blindly return `false` if you don't like what the caller gave you. This should do all these cases as `Verify.verify(...)`, i.e. this should fail! There should be a check during the construction of the QGM that makes sure that the tuple is not deeply nested, contains arrays, etc.

Reply:
in `InOpValue`, check that messages are not deep nested, or have repeated fields. and here fail if see that.